src/lib.rs

```diff
@@ -409,6 +411,14 @@ impl TlsConnectorBuilder {
         self
     }
 
+
+    #[cfg(feature = "alpn")]
+    /// Configures the set of protocols used for ALPN.
+    pub fn alpn_protocols(&mut self, protocols: &[&str]) -> &mut TlsConnectorBuilder {
```

Comment:
FYI per RFC 7301, there is no requirement for ALPNs to be UTF-8 decodable.
> Protocols are named by IANA-registered, opaque, non-empty byte strings

See also: [the list of IANA-registered ALPNs](https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids) and [RFC 8701](https://tools.ietf.org/html/rfc8701).

Reply:
But macOS's implementation of security-framework requires strings

https://docs.rs/security-framework/0.4.1/src/security_framework/secure_transport.rs.html#743-752

Reply:
Yup I agree, I think the macOS API is incorrect here. Given that both Windows and OpenSSL expose the correct APIs, IMO `rust-native-tls` should have the public API also be correct (e.g. `pub fn alpn_protocols(&mut self, protocols: &[&[u8]]) -> &mut TlsConnectorBuilder`), and probably return an `Err` when building the macOS variant if a given ALPN is not UTF-8 decodable.

Just my 2 cents from staring at this for a while, and not as anyone with authority :)

Reply:
```rust
use std::convert::TryFrom;

let mut builder = native_tls::TlsConnector::builder();

// from str (macOS/Windows/Other)
builder.alpn_protocols(&["h2", "dot", "webrtc"]);

// from bytes (Windows/Other)
let data: &[&[u8]] = &[b"h2", b"dot", b"webrtc", b"\xe2\x82\x28"];
if let Ok(alpn) = native_tls::ApplicationProtocols::try_from(data) {
    builder.alpn_protocols(alpn);
}
```